### PR TITLE
MCR-2348 create MCRXSLTransformer with a special TransformerFactory

### PR DESCRIPTION
**Pull request by rsteph-de**
[Link to jira](https://mycore.atlassian.net/browse/MCR-2348).

### CODE REVIEW

**Review by sebhofmann — APPROVED** [review]

**Comment by MyCoRe-Travis** [comment]
Something gone wrong on 285023128c3e70163fa7b58373039db67de4c277! 
 Please Check https://github.com/MyCoRe-Travis/test_artifacts/tree/6512-pull_request-1168

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java [hunk 1]
```diff
@@ -119,6 +119,11 @@ public MCRXSLTransformer() {
         super();
         setTransformerFactory(DEFAULT_FACTORY_CLASS);
     }
+    
+    public MCRXSLTransformer(Class<TransformerFactory> tfClass) {
+        super();
+        setTransformerFactory(tfClass.getName());
+    }
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: combine constructors to reuse them, e.g. link empty to this one with (DEFAULT_FACTORY_CLASS) as parameter

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java [hunk 2]
```diff
@@ -144,6 +149,18 @@ public static MCRXSLTransformer getInstance(String... stylesheets) {
...
+    
+    public static MCRXSLTransformer getInstance(Class<TransformerFactory> tfClass, String... stylesheets) {
+        String key = tfClass.getName() + "_"
+            + (stylesheets.length == 1 ? stylesheets[0] : Arrays.toString(stylesheets));
+        MCRXSLTransformer instance = INSTANCE_CACHE.get(key);
+        if (instance == null) {
+            instance = new MCRXSLTransformer(tfClass);
+            instance.setStylesheets(stylesheets);
+            INSTANCE_CACHE.put(key, instance);
+        }
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: Use same cache handling for ALL getInstance() methods, e.g. call this method from `getInstance(String... stylesheets)`

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java (hunk 2 above)
Comment (rsteph-de) [reply mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: Sebastian and I agreed on having separate cache entries for initialisation with and without transformerfactory class

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java (hunk 1 above)
Comment (rsteph-de) [reply mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: OK, done!

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java [hunk 3]
```diff
@@ -96,8 +96,9 @@
     private static long CHECK_PERIOD = MCRConfiguration2.getLong("MCR.LayoutService.LastModifiedCheckPeriod")
         .orElse(60000l);
 
-    private static final String DEFAULT_FACTORY_CLASS = MCRConfiguration2
-        .getString("MCR.LayoutService.TransformerFactoryClass").orElse(null);
+    @SuppressWarnings("unchecked")
+    private static final Class<TransformerFactory> DEFAULT_FACTORY_CLASS = (Class<TransformerFactory>) MCRConfiguration2
+        .getClass("MCR.LayoutService.TransformerFactoryClass").orElseThrow();
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: better:
```java
private static final Class<? extends TransformerFactory> DEFAULT_FACTORY_CLASS = MCRConfiguration2
    .<TransformerFactory>getClass("MCR.LayoutService.TransformerFactoryClass").orElseThrow();

```

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java [hunk 4]
```diff
@@ -169,9 +169,9 @@ public void init(String id) {
         String[] stylesheets = MCRConfiguration2.getStringOrThrow(property).split(",");
         setStylesheets(stylesheets);
         property = "MCR.ContentTransformer." + id + ".TransformerFactoryClass";
-        String transformerFactory = MCRConfiguration2.getString(property).orElse(DEFAULT_FACTORY_CLASS);
-        if (transformerFactory != null && !transformerFactory.equals(DEFAULT_FACTORY_CLASS)) {
-            setTransformerFactory(transformerFactory);
+        Optional<String> transformerFactory = MCRConfiguration2.getString(property);
+        if(transformerFactory.isPresent()) {
+            setTransformerFactory(transformerFactory.get());
```
Comment (yagee-de) [inline mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: better:
```java
MCRConfiguration2.getString("MCR.ContentTransformer." + id + ".TransformerFactoryClass")
    .ifPresent(this::setTransformerFactory);

```

**Review by yagee-de — CHANGES_REQUESTED** [review]

File: mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java (hunk 3 above)
Comment (rsteph-de) [reply mycore-base/src/main/java/org/mycore/common/content/transformer/MCRXSLTransformer.java]: "It's a kind of magic" ;-)

**Review by yagee-de — APPROVED** [review]

**Pull request merged**